types/puppeteer/index.d.ts

```diff
@@ -18,6 +18,19 @@ export type WrapElementHandle<X> = X extends Element ? ElementHandle<X> : X;
 /** Unwraps a DOM element out of an ElementHandle instance */
 export type UnwrapElementHandle<X> = X extends ElementHandle<infer E> ? E : X;
 
+export type Serializable =
+  | number
+  | string
+  | boolean
+  | null
+  | JSONArray
+  | JSONObject;
+export interface JSONArray extends Array<Serializable> { }
+export interface JSONObject {
+  [key: string]: Serializable;
+}
+export type SerializableOrJSHandle = Serializable | JSHandle;
```

Comment:
It looks like you want to catch circular references but this actually won't do that. [Here is an example.](https://www.typescriptlang.org/play/index.html#src=type%20Serializable%20%3D%0D%0A%20%20%7C%20number%0D%0A%20%20%7C%20string%0D%0A%20%20%7C%20boolean%0D%0A%20%20%7C%20null%0D%0A%20%20%7C%20JSONArray%0D%0A%20%20%7C%20JSONObject%3B%0D%0Ainterface%20JSONArray%20extends%20Array%3CSerializable%3E%20%7B%20%7D%0D%0Ainterface%20JSONObject%20%7B%0D%0A%20%20%5Bkey%3A%20string%5D%3A%20Serializable%3B%0D%0A%7D%0D%0A%0D%0Avar%20x%3A%20JSONObject%20%3D%20%7B%7D%3B%0D%0Ax%5B%22x%22%5D%20%3D%20x%3B%0D%0AJSON.stringify(x)%3B)

Reply:
Catching circular references would be great, but I'm not sure if the type system is actually powerful enough for that (aka, I don't know how to do it).

The goal of this `Serializable` is more to warn when an arg will not be correctly translated to JSON and back. For example, `RegExp`, `Date`, `Function`, etc.

The original impetus for this change was when I tried to pass an options object like `{ elem: JSHandle }` as an arg. This _contains_ a JSHandle, so I thought it would work. But it actually fails because it's neither Serializable nor a JSHandle. 

More info on exactly why JSHandles are allowed, but not nested JSHandles is in the source near this edit: https://github.com/GoogleChrome/puppeteer/pull/3591

Reply:
Ahh ok I misread the linked github thread.